File: categorical_prediction_rf.R
# this is a categorical prediction template
# You can also use library(usemodels) with their use_* to spit out templates for your model
#
# libraries
library(tidymodels)
library(palmerpenguins) # for the dataset here

## Split file into training and test set
split_file <- initial_split(penguins,strata = "species")

## FEATURE engineering ----
# Do only on training data
# also look at Feature engineering and Selection: 
# <https://bookdown.org/max/FES/>
ranger_recipe <-
  recipe(formula = species ~ ., data = training(split_file)) %>% 
    step_impute_knn(bill_length_mm, bill_depth_mm, flipper_length_mm, body_mass_g,sex,impute_with = c("island", "sex", "body_mass_g"))# impute missing values



## model specification ranger ----
ranger_spec <-
  rand_forest(trees = 1000) %>%
  set_mode("classification") %>%
  set_engine("ranger")

## put the recipe and model together into one workflow ----
ranger_workflow <-
  workflow() %>%
  add_recipe(ranger_recipe) %>%
  add_model(ranger_spec)

## fit on training set ----
trained_ranger_model <- fit(ranger_workflow,training(split_file))

# evaluate on testset ----
predicted_classes <- predict(trained_ranger_model, testing(split_file))

result <- testing(split_file) %>% add_column(predicted_classes)
## evaluating metrics
# see <https://yardstick.tidymodels.org/articles/metric-types.html>
class_metrics <- metric_set(accuracy, kap)
result %>% class_metrics(species, estimate=.pred_class)

File: categorical_prediction_rf.py
# categorical prediction with sklearn and random forest
# <https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html>
# imports ----
import numpy as np
from sklearn.datasets import make_classification # only for this example
from sklearn.ensemble import RandomForestClassifier
from sklearn.impute import SimpleImputer
from sklearn.metrics import accuracy_score, cohen_kappa_score
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline

## Split file into training and test set ----
X, y = make_classification(
    n_samples=100_000, n_features=20, n_informative=2, n_redundant=2, random_state=42
)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.33, random_state=42
)

## FEATURE engineering ----
median_imputer = SimpleImputer(missing_values=np.nan, strategy="median")
## model specification
classifier = RandomForestClassifier(max_depth=2, random_state=0)

## combine feature engineering and model
# <https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html#sklearn.pipeline.Pipeline
clf = Pipeline([("median_imputer", median_imputer), ("classification", classifier)])
## fit on trainingset
clf.fit(X_train, y_train)
## evaluate on testset
clf.score(X_test, y_test)
y_pred = clf.predict(X_test)
accuracy_score(y_test, y_pred)
cohen_kappa_score(y_test, y_pred)
